Sentiment analysis

rshar1 opened the pull request:
While the sentiment analysis is far from complete, this pull request briefly shows how I will be implementing it. I also updated the shell files so they successfully compile and run the server.

rshar1 committed:
added semantic score class

Added the class semantic score. This class contains all the methods
that will be used to keep a score updated. It will be used to connect
to the google cloud natural language API.

rshar1 committed:
outline plan and fixed formatting

Added todo comments around the code to plan changes.
Also reformatted to follow the Google Style guide.

rshar1 committed:
implemented additional methods

Implemented add message and calculate sentiment in the sentiment score
class.

Additionally, corrected the class name and added additional comments.

rshar1 committed:
corrected formatting

just some changes to formatting of multi line comments

rshar1 committed:
made some corrections to the comments

Made some corrections to the comments and code to make it more clear.

rshar1 committed:
updated sh files for external libraries

Updated the shell commands so that the application will now run with
the google natural language API

rshar1 committed:
added the third party libraries

Added the third party libraries so that the project can now compile and
run successfully

rshar1 committed:
corrected catch block

corrected the try statement by removing the finally block and correctly
placing a catch.

turtlefarmer commented on src/codeu/chat/common/SentimentScore.java:
```diff
@@ -0,0 +1,117 @@
+public class SentimentScore {
+
+  private int score;              // the average score for this object
+  private int numScores;          // this will mark how many messages have been included in the calculation of the current score
```
There's a preference for self-documenting variable names here at Google.  Private variable names are especially rarely commented.  I think score is pretty well self-defined, maybe numScores can be extended a bit to capture meaning without having to comment.

turtlefarmer commented on src/codeu/chat/common/SentimentScore.java:
```diff
@@ -0,0 +1,117 @@
+  // todo: requires environment variable GOOGLE_APPLICATION_CREDENTIALS=<PATH/TO/SERVICE-ACCOUNT/JSON>
+  /*
+   * Issue - How will this work with the google hosting. How will we make the environment variable
+   * and point to the json file. Will this be feasible?
```
Because of the hosting complication, my advice here is just to hard code it into the source code.  This is of course, terrible advice in general, but it's just not worth the effort to come up with some fancy solution to get around this artificial constraint.

rshar1 replied:
I actually wrote in this comment before I fixed up the shell files. I just submitted a separate issue describing the updated issue in more depth... #18

turtlefarmer commented on src/codeu/chat/common/SentimentScore.java:
```diff
@@ -0,0 +1,117 @@
+   * and point to the json file. Will this be feasible?
+   */
+
+  private static LanguageServiceClient languageClient;
```
static singletons pattern can cause unexpected behavior at times and also make testing more difficult.  Is there a reason why this LanguageServiceClient is static?

rshar1 replied:
When I was testing the API, I found that the first sentiment analysis would take about 10 seconds. As such, this is static so this delay only happens once, rather than for each sentiment score that is created.

turtlefarmer replied:
got it.  that's worth adding a comment for.

turtlefarmer commented on src/codeu/chat/common/SentimentScore.java:
```diff
@@ -0,0 +1,117 @@
+
+    if (m == null) return score;
+
+    Sentiment sentiment;
```
Doesn't seem necessary for this to be declared outside of the try scope.

turtlefarmer commented on src/codeu/chat/common/SentimentScore.java:
```diff
@@ -0,0 +1,117 @@
+    /*
+     * How will a specific score impact the users current score.
+     * the sentiment has both a score and a magnitude. How will they both be used in the
+     * calculation?
```
We can discuss this in our meeting.

turtlefarmer commented on src/codeu/chat/common/User.java:
```diff
@@ -52,11 +53,19 @@ public User read(InputStream in) throws IOException {
     this.id = id;
     this.name = name;
     this.creation = creation;
+    this.sentimentScore = score;
```
Does it make sense for sentiment score to be a variable into the constructor?  I'm only seeing it being passed new SentimentScore() anyways.

rshar1 replied:
When a client is given the users, they will now be given each users sentiment score, as well. This is shown in the serializer of the user. As such, this parameter is really for the client to reconstruct the users.

turtlefarmer replied:
sgtm

turtlefarmer commented:
looks good to merge.  :)

rshar1 committed:
slight changes to comments and statements

Added a comment explaining why the language client is static.

Moved a declaration inside the try block.

rshar1 committed:
Merge branch 'testing' into sentiment-analysis

# Conflicts:
#	make.sh
#	run_client.sh
#	run_server.sh
#	src/codeu/chat/client/View.java
#	test.sh

rshar1 committed:
fixed serializer in sentiment score

Fixed the serializer in the sentiment score so that it now uses gson

rshar1 committed:
added test for sentiment score serializer
